export/import: Rework how we write `migration_status.json`

PieterCK opened the pull request:
The current `get_migration_by_app` has a rather naive approach to compiling the migration status of a realm, which has led to issues like #32826. Specifically, those flaws are:
- It does not report the complete state of the migration status of the exporting servers, only the applied migration.
- It shows both the replaced and the squashing migrations. This would be a problem if we decide to delete old migrations we've replaced and import a slightly older realm with those still intact, `check_migration_status` would complain of incompatibility even though those migration files don't matter (they are replaced, after all). 
- It does not clean up ancient/stale applied migrations  (for reference, see how `check-database-compatibility` cleans those) 

This PR attempts to write a better `migration_status.json` by parsing the output of `showmigrations` instead of trying to manually write one like in `get_migration_by_app`. 

This is because Django's `showmigrations`([source](https://github.com/django/django/blob/8c118c0e00846091c261b97dbed9a5b89ceb79bf/django/core/management/commands/showmigrations.py#L113)) has a lot more logic and validations baked into it than previously thought. Ones that we care about are:
- it does validations to make sure app names are valid
- it doesn't list replaced migrations and only squashed one
- it takes into account migrations in disk(`MigrationsLoader`) vs applied migrations (`MigrationsRecorder`)

Which would resolve the first two points highlighted above.

Additionally, this PR also consolidates the list of stale migrations from `check-database-compatibility` to `lib/migration_status.py` and shares it with the migration status-related functions there to resolve the third point above.


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

PieterCK commented:
@zulipbot add "area: export/import"

andersk commented on tools/test-backend:
```diff
@@ -136,6 +136,10 @@ not_yet_fully_covered = [
     # Only covered when we run migrations tests, which are not
     # guaranteed to exist or be runnable
     "zerver/lib/migrate.py",
+    # There's a coverage-related bug for this file, see
+    # #33063 for the details and the prove that this has full
+    # test coverage.
+    "zerver/lib/migration_status.py",
```
This is because you imported `zerver.lib.migration_status` too early within `zerver.lib.test_fixtures`. The import needs to happen after is `cov.start()` is called in `tools/test-backend`. Moving the import to within `Database.what_to_do_with_migrations` should suffice.

PieterCK replied:
Ahh, I see..., works great! Thanks. #33063 will be closed after the commit moving `get_migration_status` is merged.

PieterCK commented:
Update:
- Moved import for `get_migration_status` to inside `what_to_do_with_migrations` in `test_fixtures.py`, this will resolve #33063.
- Removed [057a826](https://github.com/zulip/zulip/pull/33042/commits/057a826db0e04c81726f6b96bcbd8733abef0db3) because of the first point.

timabbott commented:
> Removed 057a826 because of the first point.

Just a note that this sort of link doesn't work in GitHub, because as soon as you remove the commit from the PR, GitHub deletes it. Generally just referencing the commit title is clear enough. (Or you can post a GitHub comment on the code itself before forcing-pushing).

PieterCK commented:
This is ready for review!

PieterCK commented:
Rebased, this is ready for review!

timabbott commented on zerver/lib/migration_status.py:
```diff
@@ -82,3 +90,39 @@ def get_migration_status(**options: Any) -> str:
+
+def parse_migration_status(
+    migration_status_print: str, stale_migrations: list[tuple[str, str]] = STALE_MIGRATIONS
+) -> AppMigrations:
```
I'm not at all convinced by the approach of parsing `showmigrations` output instead of extending `get_migrations_by_app` to make use of `MigrationsLoader` as well. It seems like it greatly reduces robustness of the system, since it means we'll be subject to the failure mode that the logic can break if `showmigrations` changes its output format in the future.

`show_list` in the Django `showmigrations.py` is only ~50 lines of code, half of which are printing logic...

But this PR fixes an unpleasant bug, and you've got some nice tests, I'm comfortable merging this version, and you can look at whether rewriting this file in the way I suggest can produce a better final implementation.

PieterCK replied:
my thinking on parsing `showmigrations` was basically just to avoid duplicating Django's codes and since we can piggyback all the validations that go in there and get what we want just from parsing its output, I'm not sure if it's worth the complexity. but in retrospect, I should have voiced this in #backend and get at least a thumbs up before implementation... :sweat_smile: 

I hadn’t weigh in the fragility of relying on parsing a print—which is not an insignificant concern. I’ll follow up by rewriting this file to mirror `show_list`.

timabbott commented:
This looks good, merged. We should be sure to manually test the Cloud export experience end-to-end after this gets deployed there.

PieterCK commented on zerver/lib/migration_status.py:
```diff
@@ -82,3 +90,39 @@ def get_migration_status(**options: Any) -> str:
+                and line[4:] in apps_stale_migrations
+            ):
+                continue
+            migrations_dict[current_app].append(line)
```
yes

cynum commented:
> This looks good, merged. We should be sure to manually test the Cloud export experience end-to-end after this gets deployed there.

I'm sorry, but doesn't seem to be working yet
https://github.com/zulip/zulip/issues/32826#issuecomment-2660869788
